src/plugins/mission/mission_impl.cpp

```diff
@@ -269,15 +269,14 @@ MissionImpl::convert_to_int_items(const std::vector<MissionItem>& mission_items)
             last_x = x;
             last_y = y;
             last_z = z;
-            last_frame = MAV_FRAME_GLOBAL_RELATIVE_ALT_INT;
 
             _mission_data.mavlink_mission_item_to_mission_item_indices.push_back(item_i);
             int_items.push_back(next_item);
         }
 
-        if (std::isfinite(item.speed_m_s)) {
-            // The speed has changed, we need to add a speed command.
-
+        // only add a speed command if the speed value is valid,
+        // the speed has changed from the previous waypoint
+        if (std::isfinite(item.speed_m_s) && (item.speed_m_s != last_speed)) {
```

Comment:
The question is how the API should be interpreted.

There are two ways:
1. NAN means no change, so don't insert the item.
2. A value means change it to that value, if it has changed in the meantime or not.

The [docstring](https://github.com/mavlink/MAVSDK/blob/1f784fec29c0a04877c6fa9b7d1213d3ac6a0a1f/src/plugins/mission/include/plugins/mission/mission.h#L92-L93) says:

> Speed to use after this mission item (in metres/second)

I can see how this is a nice optimization for someone that always sets the speed of a mission item, however, I think that's not how the API was intended to be used in the first place. You should only set it if you actually want to set it. In previous versions this was a bit clearer because a mission item used to be a class with setters/getters and so by default speed was not set, unless you would do a `set_speed(speed)`.

Reply:
Understood, I can get rid of this then.